File: src/main/java/it/pagopa/ecommerce/eventdispatcher/mdcutilities/MDCContextLifter.java
package it.pagopa.ecommerce.eventdispatcher.mdcutilities;

import java.util.AbstractMap;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import org.reactivestreams.Subscription;
import org.slf4j.MDC;
import reactor.core.CoreSubscriber;
import reactor.util.context.Context;

/**
 * Helper that copies selected values from Reactor Context to MDC for each
 * signal.
 */
class MDCContextLifter<T> implements CoreSubscriber<T> {

    CoreSubscriber<T> coreSubscriber;

    public MDCContextLifter(CoreSubscriber<T> coreSubscriber) {
        this.coreSubscriber = coreSubscriber;
    }

    @Override
    public void onSubscribe(Subscription subscription) {
        coreSubscriber.onSubscribe(subscription);
    }

    @Override
    public void onNext(T obj) {
        copyToMdc(coreSubscriber.currentContext());
        coreSubscriber.onNext(obj);
    }

    @Override
    public void onError(Throwable t) {
        try {
            copyToMdc(coreSubscriber.currentContext());
            coreSubscriber.onError(t);
        } finally {
            MDC.clear();
        }
    }

    @Override
    public void onComplete() {
        try {
            copyToMdc(coreSubscriber.currentContext());
            coreSubscriber.onComplete();
        } finally {
            MDC.clear();
        }
    }

    @Override
    public Context currentContext() {
        return coreSubscriber.currentContext();
    }

    private void copyToMdc(Context context) {
        if (!context.isEmpty()) {
            Map<String, String> mdcContextMap = Optional.ofNullable(MDC.getCopyOfContextMap()).orElseGet(HashMap::new);
            Map<String, String> reactorContextMap = Arrays
                    .stream(EventDispatcherTracingUtils.TracingEntry.values())
                    .filter(EventDispatcherTracingUtils.TracingEntry::isContextBound)
                    .map(
                            key -> new AbstractMap.SimpleEntry<>(
                                    key.getKey(),
                                    String.valueOf(
                                            context.getOrDefault(key.getKey(), key.getDefaultValue())
                                    )
                            )
                    )
                    .collect(
                            Collectors.toMap(
                                    AbstractMap.SimpleEntry::getKey,
                                    AbstractMap.SimpleEntry::getValue
                            )
                    );
            mdcContextMap.putAll(reactorContextMap);
            MDC.setContextMap(mdcContextMap);
        } else {
            MDC.clear();
        }
    }
}

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/mdcutilities/MDCContextLifter.kt
package it.pagopa.ecommerce.eventdispatcher.mdcutilities

import org.reactivestreams.Subscription
import org.slf4j.MDC
import reactor.core.CoreSubscriber
import reactor.util.context.Context

/**
 * MDC Context Lifter for transaction IDs
 */
class MDCContextLifter<T>(private val coreSubscriber: CoreSubscriber<T>) : CoreSubscriber<T> {

    companion object {
        const val TRANSACTION_ID_KEY = "transactionId"
    }

    override fun onSubscribe(subscription: Subscription) {
        coreSubscriber.onSubscribe(subscription)
    }

    override fun onNext(obj: T) {
        copyToMdc(coreSubscriber.currentContext())
        try {
            coreSubscriber.onNext(obj)
        } finally {
            MDC.clear()
        }
    }

    override fun onError(t: Throwable) {
        copyToMdc(coreSubscriber.currentContext())
        try {
            coreSubscriber.onError(t)
        } finally {
            MDC.clear()
        }
    }

    override fun onComplete() {
        copyToMdc(coreSubscriber.currentContext())
        try {
            coreSubscriber.onComplete()
        } finally {
            MDC.clear()
        }
    }

    override fun currentContext(): Context {
        return coreSubscriber.currentContext()
    }

    /**
     * Copies the transaction ID from Reactor context to MDC
     */
    private fun copyToMdc(context: Context) {
        if (!context.isEmpty) {
            context.getOrEmpty<String>(TRANSACTION_ID_KEY).ifPresent { transactionId ->
                MDC.put(TRANSACTION_ID_KEY, transactionId)
            }
        }
    }
}